### PR TITLE
[ticket/15538] Icon usage twig function

### PR DESCRIPTION
**Pull request by mrgoldy**
PHPBB3-15538

Checklist:

- [x] Correct branch: master for new features; 3.2.x for fixes
- [ ] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.2.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):
https://tracker.phpbb.com/browse/PHPBB3-15538

---
This is for the Chameleon Theme, to help out @hanakin.
No PNG icons will be shipped initially with the theme, but SVG's will be.
This is not to say that other style authors do not want to use PNG's, so there is support for png aswell.
The assumption is that icons will be located in
- `theme/icons/png/` `some_icon.png`
- `theme/icons/svg/` `some_icon.svg`

In regards to loading times, I do not know what is faster:
- hard coding html strings for the various icons and caching those
- rendering the `icon.html` through the environment

This also goes for the svg's `<path d="">`. Cause the svg's will mostlikely be _'stand-alone'_ versions of the icon (`<svg>..</svg>`) and we only want to use the the `d=""` part from the `<path>`. Anyways, I am not sure what the fastest way is to get the contents of the svg file:
- Through the environment `render()`
- Through the filesystem?
- ... ?

Moreover the svg `<path d="">`, I couldn't quite find it online yet but what are the exact allowed characters in that specific attribute? I am currently using `'/d="([a-zA-Z0-9., ]+)"/'` as a regex. Anything that needs to be included/excluded?

The current `icon.html` file needs work, was just a quick mock up to get something to be displayed.

### CODE REVIEW

**Comment by mrgoldy** [comment]
!set WIP :construction:

File: phpBB/styles/prosilver/template/icon.html [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{% if TYPE == 'png' %}
+<img class="icon png-{{ ICON ~ S_CLASSES }}" src="{{ T_PATH }}"{% if TITLE %} alt="{{ TITLE }}"{% endif %} style="background: black;">
```
Comment (hanakin) [inline phpBB/styles/prosilver/template/icon.html]: whats with the hardcoded style?

File: phpBB/styles/prosilver/template/icon.html [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+{% elseif TYPE == 'font' %}
+<i class="icon fa-{{ ICON ~ S_CLASSES }}" aria-hidden="true"{{ S_ATTRIBUTES }}></i>{% if TITLE %}<span{% if S_HIDDEN %} class="sr-only"{% endif %}>{{ TITLE }}</span>{% endif %}
```
Comment (hanakin) [inline phpBB/styles/prosilver/template/icon.html]: aria hidden should only be there if title else its should be false

File: phpBB/styles/prosilver/template/icon.html (hunk 1 above)
Comment (mrgoldy) [reply phpBB/styles/prosilver/template/icon.html]: Oh yeah my bad, I downloaded the white png 😵 couldn't see it, like I said, the entire html file needs work as it's just a mock. Decided that that will be for you.
First want to make sure the PHP and loading times are correct.

File: phpBB/styles/prosilver/template/icon.html [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+{% elseif TYPE == 'svg' %}
+<svg class="icon svg-{{ ICON ~ S_CLASSES }}"{% if TITLE %} aria-hidden="true" aria-labeled-by="{{ TITLE_CLEAN }}"{% endif %} xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"  width="24" height="24" viewBox="0 0 24 24">
```
Comment (hanakin) [inline phpBB/styles/prosilver/template/icon.html]: we do not need all of the extra stuff here you can remove
```
xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"  width="24" height="24"
```

**Comment by hanakin** [comment]
@CHItA @Nicofuma you insights for the best way to implement this would probably help as well. IIRC one of you had an idea for handling this

**Comment by hanakin** [comment]
have a look at some of the code from this old pr https://github.com/phpbb/phpbb/pull/4765/files might help with testing and what not

**Comment by hanakin** [comment]
@mrgoldy where did we leave of with this? should probably create some tests. We will need to test it in all locations/possible uses that is to say I want to use it from icons, imagesets, contact icons, forum icons, smiles

the last one is a maybe

but this means we need to be able to specify in the case of forum icons a different path as they are in the files folder same with the smiles.

**Pull request closed without merging**